app/channel.go

```diff
@@ -45,71 +45,54 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
 		}
 	}
 
-	if result := <-a.Srv.Store.Channel().GetByName(teamId, "town-square", true); result.Err != nil {
-		err = result.Err
-	} else {
-		townSquare := result.Data.(*model.Channel)
-
-		cm := &model.ChannelMember{
-			ChannelId:   townSquare.Id,
-			UserId:      user.Id,
-			SchemeUser:  true,
-			SchemeAdmin: shouldBeAdmin,
-			NotifyProps: model.GetDefaultChannelNotifyProps(),
-		}
-
-		if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
-			err = cmResult.Err
-		}
-		if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, townSquare.Id, model.GetMillis()); result.Err != nil {
-			mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
-		}
-
-		if *a.Config().ServiceSettings.ExperimentalEnableDefaultChannelLeaveJoinMessages {
-			if requestor == nil {
-				if err := a.postJoinTeamMessage(user, townSquare); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			} else {
-				if err := a.postAddToTeamMessage(requestor, user, townSquare, ""); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			}
-		}
+	defaultChannelList := []string{"town-square"}
 
-		a.InvalidateCacheForChannelMembers(result.Data.(*model.Channel).Id)
+	if len(a.Config().TeamSettings.ExperimentalDefaultChannels) == 0 {
+		defaultChannelList = append(defaultChannelList, "off-topic")
+	} else {
+		defaultChannelList = append(defaultChannelList, a.Config().TeamSettings.ExperimentalDefaultChannels...)
 	}
 
-	if result := <-a.Srv.Store.Channel().GetByName(teamId, "off-topic", true); result.Err != nil {
-		err = result.Err
-	} else if offTopic := result.Data.(*model.Channel); offTopic.Type == model.CHANNEL_OPEN {
+	for _, channelName := range defaultChannelList {
```

Comment:
Do we need to do any duplicate checking here? What happens if someone sets the setting to `["town-square", "town-square", "town-square", "town-square"]`?

Reply:
Good catch. Duplicate checking makes sense, I don't think anything particular bad would happen, but it's certainly not harmful.